src/virtual_dom/patch/patch_gen.rs

```diff
@@ -0,0 +1,474 @@
+                    self.yield_keyed()
+                } else {
+                    let next_node = child_old
+                        .node_ws()
+                        .expect("Old node not connected to web_sys node")
+                        .clone();
+                    self.insert(child_new, next_node)
+                }
+            }
+            (None, Some(_)) => {
+                let child_old = self
+                    .old_children
+                    .pop_back()
+                    .expect("No node with a matching key");
+                if make_key(&child_old) == self.matching_key {
+                    self.matching_child_old.replace(child_old);
+                    self.yield_keyed()
+                } else {
+                    self.remove(child_old)
+                }
+            }
+            (None, None) => {
+                if let (Some(child_old), Some(child_new)) =
+                    (self.old_children.pop_back(), self.new_children.pop_back())
+                {
+                    if make_key(&child_old) == self.matching_key
+                        && make_key(child_new) == self.matching_key
+                    {
+                        self.matching_child_old.replace(child_old);
+                        self.matching_child_new.replace(child_new);
+                        self.yield_keyed()
+                    } else if make_key(&child_old) == self.matching_key {
+                        let next_node = child_old.node_ws().unwrap().clone();
+                        self.matching_child_old.replace(child_old);
+                        self.insert(child_new, next_node)
+                    } else if make_key(child_new) == self.matching_key {
+                        self.matching_child_new.replace(child_new);
+                        self.remove(child_old)
+                    } else {
+                        self.patch_or_replace(child_old, child_new, true)
+                    }
+                } else {
+                    // if the matching key is set then both the old and new children queues has
+                    // node with this key.
+                    unreachable!("No node with a matching key")
+                }
+            }
+        }
+    }
+
+    fn yield_remaining(&mut self) -> Option<PatchCommand<'a, Ms>> {
+        let maybe_child_old = self.old_children.pop_back();
+        let maybe_child_new = self.new_children.pop_back();
+
+        match (maybe_child_old, maybe_child_new) {
+            (Some(child_old), Some(child_new)) => self.patch_or_replace(child_old, child_new, true),
+            (Some(child_old), None) => self.remove(child_old),
+            (None, Some(child_new)) => self.append(child_new),
+            (None, None) => None,
+        }
+    }
+
+    fn next_node_ws(&mut self) -> Option<web_sys::Node> {
+        self.old_children
+            .iter()
+            .find_map(|old| old.node_ws().cloned())
+            .or_else(|| loop {
```

Comment:
Loops are mostly used as `Iterator::find()` to consume iterators by calling `next()`, filling the internal buffer with the new items or returning that item if it meets the condition.

Reply:
Well, I would much rather see that `Iterator::find`. But I assume you have good reasons for it - please add at least some comment into it.

Reply:
This code could be replaced by the `Itertools::zip_longest` if we could add the new dependence.
```rust
    fn yield_keyless(&mut self) -> Option<PatchCommand<'a, Ms>> {
        // Take a pair of old/new children but skip if both are `Some(Node::Empty)`.
        let (child_old, child_new) = loop {
            // First consume the children stored in the queue.
            // When old child is `Empty` we call `find_next_node_ws` which
            // moves some children from the source iterator to the queue.
            let old = self
                .old_children
                .pop_back()
                .or_else(|| self.old_children_iter.next());
            let new = self.new_children_iter.next();
            // We should not issue any command if both the old and the new nodes are `Empty`.
            if let (Some(Node::Empty), Some(Node::Empty)) = (&old, &new) {
                continue;
            }
            break (old, new);
        };
```

Reply:
My bad, it could not. Because the `zip_longest` consumes the iterators by value.

Reply:
Don't be afraid to add new dependencies if it makes code more readable.
Exceptions
  1. Some dependencies are too big - e.g. we can't use `url` (and everything that uses it like `reqwest`) because `url` contains unicode tables and other things that double the WASM file size...
  2. Dependencies that contain entire subsystems (e.g. VirtualDOM, template engine, etc.) because the most of them are in alpha/beta state or they are experiments and we will lose some flexibility.

And yeah, that code snippet looks much better.

Reply:
> My bad, it could not. Because the `zip_longest` consumes the iterators by value.

Yeah, it was another reason why I recommended to use pure functions and use immutable values or owned ones - I was fighting with iterators, streams and other things until I realized it and adapted my writing style a bit.